### PR TITLE
feature(bulk): add bulk insert with callback for scaning returning id

### PR DESCRIPTION
**Pull request by laushunyu**
feature: allow user to scan `returning id` after every bulk insert

### CODE REVIEW

**Comment by t-tiger** [comment]
Thank you for your contribution. I'm reviewing and considering the interface to meet your demand now.
I think it'll finish within a few days, sorry for letting you wait for a while.

**Comment by t-tiger** [comment]
@laushunyu 

I'm thinking of 2 ways to achieve your demand.

1. Add a function to return *gorm.DB 

```golang
func BulkInsertReturningDB(db *gorm.DB, objects []interface{}, chunkSize int, excludeColumns ...string) (*gorm.DB, error)
```

2. Add a dedicated function to return inserted IDs

```golang
func BulkInsertReturningIDs(db *gorm.DB, objects []interface{}, chunkSize int, excludeColumns ...string) ([]int, error)
```

At first, I was about to take option "1" since it's open to various use cases and callers can utilize it to meet their demands.

But now, I'm thinking option "2" is better. "2" is not for multiple use cases, but the point is, the interface is clear and callers easily expect what will happen with this function. If there are lots of use cases for bulk insert, option "1" may be good, but there were not many clear use cases compared to your demand. So, I want to add a dedicated function.

What do you think of this? If you agree, I'll create a PR and want you to review it.

**Comment by laushunyu** [comment]
option "2" is not fitted to get returning uuid. option "1" is better i think, we can append err to `gorm.DB` then return gorm.DB only(like gorm), but i dont know how to merge `gorm.DB`s splited by `chunkSize`.

**Comment by t-tiger** [comment]
@laushunyu 

Sorry for the late response and thank you for the opinion.

> i dont know how to merge gorm.DBs splited by chunkSize.

That's a good point! And I considered this,

> option "2" is not fitted to get returning uuid

I figured out the new interface of the following.

```golang
func BulkInsertWithAssigningIDs(db *gorm.DB, ids interface{}, objects []interface{}, chunkSize int, excludeColumns ...string) error
```

And here is the caller.

```golang
// In case of UUID
var uuids []string
if err := gormbulk.BulkInsertWithAssigningIDs(db, &uuids, as, 1000); err != nil {
	panic(err)
}

// In case of int
var ids []int
if err := gormbulk.BulkInsertWithAssigningIDs(db, &ids, as, 1000); err != nil {
	panic(err)
}
```

We can deal with both `[]string` and `[]int` by using the reflection, and assign the entire ids even if the values are split with chunks. I feel like it's a better solution, what do you think?

**Comment by laushunyu** [comment]
@t-tiger 
Great, so far this is the best solution.I'm looking forward to this feature~

**Comment by laushunyu** [comment]
@t-tiger 
i implement the interface suggested, would u like to review it again?

**Comment by laushunyu** [comment]
@t-tiger 
additionally, i think the inteface `BulkInsertWithAssigningIDs(db *gorm.DB, ids interface{}, objects []interface{}, chunkSize int, excludeColumns ...string) error` can also handle `returning *` when i implement it😆😆😆

File: README.md
Comment (t-tiger) [inline README.md:99]: I appreciate your cooperation to update README, but can I have a chance to change README as a follow-up task after merging? I actually want to review entire sentences.

File: bulk_insert.go
Comment (t-tiger) [inline bulk_insert.go]: As far as I tested, the current one does not affect the value of `returningId`.

```suggestion
	allIDs := reflect.Indirect(reflect.ValueOf(returningId))
```

File: bulk_insert.go
Comment (t-tiger) [inline bulk_insert.go]: I think "returning" is a bit confusing for the caller since "returning" comes from DB side, and just defining `ids interface{}` is better IMO.

File: bulk_insert.go
Comment (t-tiger) [inline bulk_insert.go]: `Scan` does not work as expected. I think we should use `Pluck` instead.

However, at the same time, we have to know the primary key of objects since not all table's primary key is named "id", so we need to examine `*gorm.Field` to confirm what is the primary key.

**Comment by t-tiger** [comment]
@laushunyu 
I left some review comments, and I'm ready to add a commit to fix them.

By the way, I've created an epic branch of "epic-aggining-ids" and I'm considering that we'll merge this PR to epic branch as it is and add follow-up PRs to fix with my comments. After finishing the follow-ups, merge an epic branch to master. Is it OK to proceed with this plan?

File: bulk_insert.go
Comment (bombsimon) [inline bulk_insert.go:159]: Why do we change from `Exec` to `Raw`?

File: bulk_insert_test.go
Comment (bombsimon) [inline bulk_insert_test.go:121]: The test case this was coped from was to test the `gorm:"column"` tag. I think it makes the test unnecessary complex to not just use a simple tags and names as possible.

File: bulk_insert_test.go
Comment (bombsimon) [inline bulk_insert_test.go:128]: The feature you're testing is PostgreSQL specific so it would make more sense to use the postgres dialect here even though it's just a mock.

File: bulk_insert_test.go
Comment (bombsimon) [inline bulk_insert_test.go:160]: If you're going to add a public method that will basically do this for you I think it would make more sense to test the public interface. This way we actually test what's exposed to the users and at the same time the logic you're testing here.

If you just want to ensure a defined callback is called I would simplify the test without additional plucks and just assert some variable is incremented or whatnto.

File: bulk_insert_test.go
Comment (bombsimon) [inline bulk_insert_test.go:166]: This is basically the same thing twice, the test will fail if there's an error. 

```suggestion
	require.NoError(t, err)
```

**Review by bombsimon — COMMENTED** [review]
Nice PR! ⭐ 

I think the original example in the README and the callback is nice but `RETURNING id` is a PostgreSQL specific feature so I'm not sure it makes sense to add that as a public method in this library. If we do, maybe it would be and idea to create separate packages for each dialect not only to make some separation of concerns but also to make it more clear to the user they can't use this feature for MySQL.

`ON DUPLICATE UPDATE` has been requested a bunch of times which is easy enough to implement with insert options. That's however MySQL specific so we haven't added it (yet). With a separate MySQL package, maybe that could be added as well?

File: bulk_insert.go
Comment (laushunyu) [reply bulk_insert.go]: yes, my playground is 
```
package main

import (
	"fmt"
	"time"

	"github.com/jinzhu/gorm"
	_ "github.com/jinzhu/gorm/dialects/postgres"
	gormbulk "github.com/t-tiger/gorm-bulk-insert"
)

type Model struct {
	Id        uint      `gorm:"primary_key"`
	CreatedAt time.Time `gorm:"default:now()"`
	Data      string
}

func main() {
	db, err := gorm.Open("postgres", "host=localhost port=5432 user=cloudwalker dbname=cloudwalker password=cloudwalker sslmode=disable")
	if err != nil {
		panic(err)
	}
	defer db.Close()
	db.SingularTable(true)

	if err := db.AutoMigrate(Model{}).Error; err != nil {
		panic(err)
	}

	models := []interface{}{
		Model{Data: "aaa"},
		Model{Data: "bbb"},
		Model{Data: "ccc"},
	}
	var returnModel []Model
	if err := gormbulk.BulkInsertWithAssigningIDs(
		db.Set("gorm:insert_option", "returning *"), &returnModel, models, 1000); err != nil {
		panic(err)
	}

	fmt.Printf("insert success with returning: %#v", returnModel)
}
```
it works well, but i miss a little bit of details.
IMO, we can assert ids(interface{}) as
- []struct: Scan ids
- []T: Pluck "primary key"

what do u think of?

File: bulk_insert.go
Comment (laushunyu) [reply bulk_insert.go:159]: I find a commentary in gorm for `Raw`: 
```
// Raw use raw sql as conditions, won't run it unless invoked by other methods
//    db.Raw("SELECT name, age FROM users WHERE name = ?", 3).Scan(&result)
```
We can `Scan` or `Pluck` the result after `Raw` not `Exec`

File: bulk_insert_test.go
Comment (laushunyu) [reply bulk_insert_test.go:160]: yeap, i will change it

File: bulk_insert.go
Comment (laushunyu) [reply bulk_insert.go]: good idea

File: bulk_insert.go
Comment (laushunyu) [reply bulk_insert.go]:
```
// it will scan the result of `returning id` or `returning *` to [returnedValue] after every insert.
// it's necessary to set "gorm:insert_option"="returning id" in *gorm.DB
// [returnedValue] slice of primary_key or model, must be a *[]uint(for integer), *[]string(for uuid), *[]struct(for `returning *`)
```
It can be more general in some case.  we can scan what we inserted using `returning *` back to the value, so i renamed it as `returnedValue` (i write more example in README). what do u think of it?

File: bulk_insert_test.go
Comment (laushunyu) [reply bulk_insert_test.go:121]: Sorry, i have no idea how to test it with `github.com/DATA-DOG/go-sqlmock`, i never use this packge. but i know that it works with my postgres database. would u plz help me for testing it after merging into epic branch?

**Comment by laushunyu** [comment]
@t-tiger  @bombsimon 
Thanks for revewing. it's a good idea to merge it into epic branch, i need more help for writing test code of new feature.

File: bulk_insert_test.go
Comment (bombsimon) [reply bulk_insert_test.go:121]: Sure thing, I can write tests and push to this branch before merging as well, that's probably better. I'll hold off doing that until @t-tiger approves this PR so we know it's the final draft.

**Comment by t-tiger** [comment]
I'll merge to epic branch.

**Review by t-tiger — APPROVED** [review]

File: bulk_insert.go
Comment (t-tiger) [reply bulk_insert.go]: Yeah, we can assign struct values here. But I think we should separate methods for the case of receiving slice of struct and slice of IDs. Please take care of it next PR.

**Pull request merged**
